### ISSUE TITLE
Add the option to process all files in a directory in force-cube

### ISSUE THREAD

**Issue by maawoo** [davidfrantz/force#63]
Hi there! 

As far as I know, `force-cube` can only process single files.
It would be great to have the option to run `force-cube` over all files (e.g. Sentinel-1 GeoTIFF files) in a directory. In the meantime I'll just write my own script to do that, but to have the option with the `force-cube`-program probably would be helpful for others as well!

Cheers, Marco

**Comment by ernstste** [comment]
Hi maawoo,

You can process all files within a directory by using commands like `find` with the `-exec` option or even just a for loop with one command line call.

Stefan

**Comment by davidfrantz** [comment]
Hi Marco,

There are several ways to do this without the need of a script (potentially this might be included in ``force-cube`` in the future, too).

Here are some options (assuming you are in the correct directory):

1) ``find . -name '*.tif' -exec force-cube {} datacube-dir cubic 10``

2) ``for f in *.tif; do force-cube $f datacube-dir cubic 10; done``

3) ``ls *.tif | parallel -j 1 force-cube {} datacube-dir cubic 10``

I personally use 

1) if input files are in different subdirectories (``find`` is recursive)

2) if input files are in one directory

3) if images are in one directory, and I need an ETA (just use ``--eta`` as additional option for ``parallel``)


Whatever you do, *please consider the following*:

- ``force-cube`` already has some built-in parallelization. If your input image is falling into several output tiles, you will get a parallel job for each tile!
- input images with the same basename will be merged into the same output image if they fall into the same output tile (e.g. ``WRS2-170-072/20150701_LEVEL2_LND07_BOA.tif`` and ``WRS2-170-073/20150701_LEVEL2_LND07_BOA.tif`` will have an overlap area that will be merged into one file)

This has two important implications:

- you really need to take care when calling ``force-cube`` from a loop, especially if the loop is parallel. If properly done, this is essentially possible, but a naive implementation risks race conditions (note that I used ``-j 1`` in the third option above to make the calling loop sequential)
- you might exhaust the number of possible threads on your system when using a parallel loop. This is because this will result in a nested parallelization. This can especially happen when using many tasks, or when input images are large compared to the output tiles.

I hope this helps a bit, cheers
David

**Comment by ernstste** [comment]
Great answer, David. 
Just a heads up that the find command needs to be terminated using `\;` when used with the -exec option.

**Comment by davidfrantz** [comment]
Thanks for noticing, Stefan,
I've edited my answer above :)

**Comment by maawoo** [comment]
Thank you both for your answers and especially David for taking your time to go a bit more in-depth! 🙂

**Comment by davidfrantz** [comment]
@maawoo ,

push https://github.com/davidfrantz/force/commit/fb2c0a9a6fd210b06b38452be084cb3d6b03e6d0 in the develop branch adds the new possibility to cube multiple input files, e.g. like this:

`force-cube *`
`force-cube classification.tif aoi.gpkg images_*.tif`


Note that the input files are cubed one after another (no parallelization), wherein each image is cubed in parallel (like I described above with option 2).

Cheers,
David
